lib/tom_queue/worker_supervisor.rb

```diff
@@ -0,0 +1,67 @@
+module TomQueue
+# class WorkerPool?
+  class WorkerSupervisor
+
+    attr_reader :processes
+
+    def initialize(processes = default_processes)
+      @processes = processes
+      setup_signal_handling
+    end
+
+    def default_processes
+      # TODO: deferred worker
+      {
+        "worker1" => -> { TomQueue::Worker.new.start },
+        "worker2" => -> { TomQueue::Worker.new.start },
+        "worker3" => -> { TomQueue::Worker.new.start },
+        "worker4" => -> { TomQueue::Worker.new.start },
+        "scheduler" => -> { TomQueue::Worker.new.start }
+      }
+    end
+
+    def setup_signal_handling
+      ["SIGINT","SIGTERM"].each do |signal|
+        Signal.trap(signal) do
+          raise SignalException, "SIGTERM"
+        end
+      end
+    end
+
+    def start_workers
+      pids = {}
+      # TODO: before_fork
+      loop do
+        processes_to_start = (processes.keys - pids.values)
+        processes_to_start.each do |name|
+          $stderr.puts "Starting '#{name}'..."
+          pid = fork_worker(processes[name])
+          pids[pid] = name
+        end
+
+        pid, status = wait_for_child_process_to_exit
+
+        if name = pids.delete(pid)
+          $stderr.puts "#{name} process terminated (#{pid}) with status #{status}"
+        end
+        sleep 1
+      end
+    end
+
+    private
+
+    def fork_worker(process)
+      # TODO: What about the before_fork here instead?
+      fork do
+        # TODO: after_fork
+        process.call
+        # what's the exit(0) doing here? Saying "We completed OK" when the worker shutsdown?
+        exit(0)
```

Comment:
Yeah - I can never remember if the block "falls-through" so carries on executing the parent process code once it's finished processing the code in the child. This way it's unambiguous.

I'd also be tempted to wrap it in a begin/ensure block to ensure exceptions don't accidentally leak out of the fork block.